src/sync/SegmentUpdateWorker/SegmentUpdateWorker.js

```diff
@@ -0,0 +1,50 @@
+/**
+ * SegmentUpdateWorker class
+ */
+export default class SegmentUpdateWorker {
+
+  /**
+   * @param {Object} segmentsStorage
+   * @param {Object} segmentsProducer
+   */
+  constructor(segmentsStorage, segmentsProducer) {
+    this.segmentsStorage = segmentsStorage;
+    this.segmentsProducer = segmentsProducer;
+    this.segmentsChangesQueue = [];
+  }
+
+  // Private method
+  // Preconditions: this.segmentsProducer.isSynchronizeSegmentRunning === false
+  __handleSegmentUpdateCall() {
+    if (this.segmentsChangesQueue.length > 0) {
+      const { changeNumber, segmentName } = this.segmentsChangesQueue[this.segmentsChangesQueue.length - 1];
+      if (changeNumber > this.segmentsStorage.getChangeNumber(segmentName)) {
+        this.segmentsProducer.synchronizeSegment(segmentName).then(() => {
+          this.__handleSegmentUpdateCall();
+        });
+      } else {
+        this.segmentsChangesQueue.pop();
+        this.__handleSegmentUpdateCall();
+      }
+    }
+  }
+
+  /**
+   * Invoked by NotificationProcessor on SEGMENT_UPDATE event
+   *
+   * @param {number} changeNumber change number of the SEGMENT_UPDATE notification
+   * @param {string} segmentName segment name of the SEGMENT_UPDATE notification
+   */
+  put(changeNumber, segmentName) {
+    const currentChangeNumber = this.segmentsStorage.getChangeNumber(segmentName);
+
+    if (changeNumber <= currentChangeNumber) return;
+
+    this.segmentsChangesQueue.push({ segmentName, changeNumber });
+
+    if (this.segmentsProducer.isSynchronizeSegmentRunning()) return;
```

Comment:
This is odd. You can sync few segments in parallel and there's no reason not to update segment A at the same time than segment B, it just doesn't make sense to trigger two requests at the same time for segment B.

Reply:
Done. Now `SegmentUpdateWorker` keeps a map of segment names to `maxChangeNumbers`, instead of a `segmentsChangesQueue` (similar approach followed by `SplitUpdateWorker` and `MySegmentUpdateWorker`)